aea/helpers/dialogue/base.py

```diff
@@ -437,65 +391,67 @@ def is_empty(self) -> bool:
         :return: True if message successfully added, false otherwise
         """
-        if not message.is_incoming:
+        if not message.has_sender:
             message.sender = self.agent_address
 
-        self.ensure_counterparty(message)
-
         if not self.is_belonging_to_dialogue(message):
-            return False
-
-        is_extendable = self.is_valid_next_message(message)
-        if is_extendable:
-            if message.is_incoming:
-                self._incoming_messages.extend([message])
-            else:
-                self._outgoing_messages.extend([message])
-        return is_extendable
+            raise InvalidDialogueMessage("message does not belong to this dialogue.")
 
-    def ensure_counterparty(self, message: Message) -> None:
-        """
-        Ensure the counterparty is set (set if not) correctly.
+        is_valid_result, is_valid_message = self.is_valid_next_message(message)
 
-        :param message: a message
-        :return: None
-        """
-        counterparty = None  # type: Optional[str]
-        try:
-            counterparty = message.counterparty
-        except AssertionError:
-            # assume message belongs to dialogue
-            message.counterparty = self.dialogue_label.dialogue_opponent_addr
+        if not is_valid_result:
+            raise InvalidDialogueMessage(is_valid_message)
 
-        if counterparty is not None:
-            assert (
-                message.counterparty == self.dialogue_label.dialogue_opponent_addr
-            ), "The counterparty specified in the message is different from the opponent in this dialogue."
+        if self.is_message_by_self(message):
+            self._outgoing_messages.extend([message])
+        else:
+            self._incoming_messages.extend([message])
 
     def is_belonging_to_dialogue(self, message: Message) -> bool:
         """
         Check if the message is belonging to the dialogue.
 
         :param message: the message
-        :return: Ture if message is part of the dialogue, False otherwise
+        :return: True if message is part of the dialogue, False otherwise
         """
         if self.is_self_initiated:
+            if self.is_message_by_self(message):
+                opponent = message.to
+            else:
+                opponent = message.sender
             self_initiated_dialogue_label = DialogueLabel(
-                (message.dialogue_reference[0], Dialogue.OPPONENT_STARTER_REFERENCE),
-                message.counterparty,
+                (
+                    message.dialogue_reference[0],
+                    Dialogue.UNASSIGNED_DIALOGUE_REFERENCE,
+                ),
+                opponent,
                 self.agent_address,
             )
             result = self_initiated_dialogue_label in self.dialogue_labels
         else:
+            if self.is_message_by_self(message):
+                opponent = message.to
+            else:
+                opponent = message.sender
```

Comment:
one liner possible...also since you have a need to get the opponent in multiple places; why not expose it as a property on the message?